File: examples/leetcode-out/zig/18/four-sum.zig
const std = @import("std");

fn expect(cond: bool) void {
    if (!cond) @panic("expect failed");
}

fn fourSum(nums: []const i32, target: i32) []const []const i32 {
    const sorted: []const i32 = blk: { var _tmp0 = std.ArrayList(struct { item: i32; key: i32 }).init(std.heap.page_allocator); for (nums) |n| { _tmp0.append(.{ .item = n, .key = n }) catch unreachable; } for (0.._tmp0.items.len) |i| { for (i+1.._tmp0.items.len) |j| { if (_tmp0.items[j].key < _tmp0.items[i].key) { const t = _tmp0.items[i]; _tmp0.items[i] = _tmp0.items[j]; _tmp0.items[j] = t; } } } var _tmp1 = std.ArrayList(i32).init(std.heap.page_allocator);for (_tmp0.items) |p| { _tmp1.append(p.item) catch unreachable; } var _tmp2 = _tmp1.toOwnedSlice() catch unreachable; break :blk _tmp2; };
    const n: i32 = (sorted).len;
    var result = std.ArrayList(i32).init(std.heap.page_allocator);
    for (@as(i32,@intCast(0)) .. n) |i| {
        if ((((i > @as(i32,@intCast(0))) and sorted[i]) == sorted[(i - @as(i32,@intCast(1)))])) {
            continue;
        }
        for ((i + @as(i32,@intCast(1))) .. n) |j| {
            if (((((j > i) + @as(i32,@intCast(1))) and sorted[j]) == sorted[(j - @as(i32,@intCast(1)))])) {
                continue;
            }
            var left: i32 = (j + @as(i32,@intCast(1)));
            var right: i32 = (n - @as(i32,@intCast(1)));
            while ((left < right)) {
                const sum: i32 = (((sorted[i] + sorted[j]) + sorted[left]) + sorted[right]);
                if ((sum == target)) {
                    try result.append(@as(i32,@intCast(&[_]i32{sorted[i], sorted[j], sorted[left], sorted[right]})));
                    left = (left + @as(i32,@intCast(1)));
                    right = (right - @as(i32,@intCast(1)));
                    while ((((left < right) and sorted[left]) == sorted[(left - @as(i32,@intCast(1)))])) {
                        left = (left + @as(i32,@intCast(1)));
                    }
                    while ((((left < right) and sorted[right]) == sorted[(right + @as(i32,@intCast(1)))])) {
                        right = (right - @as(i32,@intCast(1)));
                    }
                } else                 if ((sum < target)) {
                    left = (left + @as(i32,@intCast(1)));
                } else {
                    right = (right - @as(i32,@intCast(1)));
                }
            }
        }
    }
    return result.items;
}

fn test_example_1() void {
    expect((fourSum(&[_]i32{@as(i32,@intCast(1)), @as(i32,@intCast(0)), -@as(i32,@intCast(1)), @as(i32,@intCast(0)), -@as(i32,@intCast(2)), @as(i32,@intCast(2))}, @as(i32,@intCast(0))) == &[_][]const i32{&[_]i32{-@as(i32,@intCast(2)), -@as(i32,@intCast(1)), @as(i32,@intCast(1)), @as(i32,@intCast(2))}, &[_]i32{-@as(i32,@intCast(2)), @as(i32,@intCast(0)), @as(i32,@intCast(0)), @as(i32,@intCast(2))}, &[_]i32{-@as(i32,@intCast(1)), @as(i32,@intCast(0)), @as(i32,@intCast(0)), @as(i32,@intCast(1))}}));
}

fn test_example_2() void {
    expect((fourSum(&[_]i32{@as(i32,@intCast(2)), @as(i32,@intCast(2)), @as(i32,@intCast(2)), @as(i32,@intCast(2)), @as(i32,@intCast(2))}, @as(i32,@intCast(8))) == &[_][]const i32{&[_]i32{@as(i32,@intCast(2)), @as(i32,@intCast(2)), @as(i32,@intCast(2)), @as(i32,@intCast(2))}}));
}

pub fn main() void {
    test_example_1();
    test_example_2();
}

File: examples/leetcode-out/swift/18/four-sum.swift
import Foundation

func expect(_ cond: Bool) {
  if !cond { fatalError("expect failed") }
}

func _index<T>(_ arr: [T], _ i: Int) -> T {
  var idx = i
  let n = arr.count
  if idx < 0 { idx += n }
  if idx < 0 || idx >= n { fatalError("index out of range") }
  return arr[idx]
}

func fourSum(_ nums: [Int], _ target: Int) -> [[Int]] {
  let nums = nums
  let target = target

  let sorted: [Int] = nums.sorted()
  let n = sorted.count
  var result: [[Int]] = []
  for i in 0..<n {
    if i > 0 && _index(sorted, i) == _index(sorted, i - 1) {
      continue
    }
    for j in i + 1..<n {
      if j > i + 1 && _index(sorted, j) == _index(sorted, j - 1) {
        continue
      }
      var left = j + 1
      var right = n - 1
      while left < right {
        let sum =
          _index(sorted, i) + _index(sorted, j) + _index(sorted, left) + _index(sorted, right)
        if sum == target {
          result =
            result + [
              [_index(sorted, i), _index(sorted, j), _index(sorted, left), _index(sorted, right)]
            ]
          left = left + 1
          right = right - 1
          while left < right && _index(sorted, left) == _index(sorted, left - 1) {
            left = left + 1
          }
          while left < right && _index(sorted, right) == _index(sorted, right + 1) {
            right = right - 1
          }
        } else if sum < target {
          left = left + 1
        } else {
          right = right - 1
        }
      }
    }
  }
  return result
}

func test_example_1() {
  expect(fourSum([1, 0, -1, 0, -2, 2], 0) == [[-2, -1, 1, 2], [-2, 0, 0, 2], [-1, 0, 0, 1]])
}

func test_example_2() {
  expect(fourSum([2, 2, 2, 2, 2], 8) == [[2, 2, 2, 2]])
}

func main() {
  test_example_1()
  test_example_2()
}
main()
